src/basic/strv.c

```diff
@@ -353,6 +353,55 @@ int strv_split_extract(char ***t, const char *s, const char *separators, Extract
         return (int) n;
 }
 
+int strv_split_extract_pairs_colon(char ***t, const char *s) {
+        _cleanup_strv_free_ char **l = NULL;
+        size_t n = 0, allocated = 0;
+        int r;
+
+        assert(t);
+        assert(s);
+
+        for (;;) {
+                _cleanup_free_ char *first = NULL, *second = NULL, *tuple = NULL, *second_or_empty = NULL;
+                const char *p = NULL;
+
+                r = extract_first_word(&s, &tuple, WHITESPACE, EXTRACT_UNQUOTE|EXTRACT_DONT_COALESCE_SEPARATORS);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        break;
+
+                p = tuple;
+                r = extract_many_words(&p, ":", EXTRACT_UNQUOTE|EXTRACT_DONT_COALESCE_SEPARATORS, &first, &second, NULL);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        continue;
```

Comment:
this will parse out two words, but if there are actually three colon separated words, then you ignore the last one... you need to check here that `isempty(p)` holds after you are done

Reply:
Ok, changed to fail explicitly with -EINVAL instead of discarding extra words